Don't suggest a semicolon when one already exists

sloanev opened the pull request:
In cases where it can be confirmed that an expression with borrow conflicts is already followed by an expression, the compiler will no longer suggest adding a second semicolon.

This is my first time contributing to the compiler; I eagerly accept any feedback on improvements or how this could be done differently.

closes #133941

Commit:
Don't suggest a semicolon when one already exists

lcnr reviewed (commented):
Hi :wave: please add a regression test to an appropriate subfolder of `tests/ui`

lcnr commented on compiler/rustc_borrowck/src/diagnostics/explain_borrow.rs:
```diff
@@ -248,11 +249,27 @@ impl<'tcx> BorrowExplanation<'tcx> {
+                                    match source_map.span_to_snippet(span) {
+                                        Ok(span_str) => !span_str.contains(';'),
+                                        Err(_) => true,
+                                    }
```
we probably want to change `BlockTailInfo::tail_result_is_ignored` instead. Using spans for such checks does not work in case the error involves macro calls. The current check may also have issues if something like `tail_expr ;` or `tail_expr/* */;` or `tail_expr<new-line>;` is used

I could imagine that we should either change `tail_result_is_ignored` to an enum tracking why it's ignored or alternatively change `LocalInfo::BlockTailTemp` to track whether the tail is an expr or a statement

joculatrix replied:
That makes sense! Thanks for the suggestion - I knew someone who knew more than me would steer me in a better direction. I'll look into that. 👍

Commit:
Add regression test

joculatrix commented:
@lcnr Sorry if you're not a good person to ask -- as far as I can tell, information about whether or not a statement/block ends with a semicolon is lost when `hir::StmtKind::Expr` and `hir::StmtKind::Semi` both get lowered to the same representation in the THIR, which then doesn't pass it on to the MIR. The error message in question is caught during borrow-checking, which uses the MIR. I just want to ask before doing it because I'm a new contributor and don't want to go beyond the scope of this issue/PR: is it appropriate to go back as far as editing how the HIR is lowered to THIR and how that then is lowered to MIR in order to preserve whether statements end with semicolons? Or, if not, is there some other way of getting this information that I'm missing?

lcnr commented:
> : is it appropriate to go back as far as editing how the HIR is lowered to THIR and how that then is lowered to MIR in order to preserve whether statements end with semicolons?

If we currently end up erasing that information, making sure it's tracked somewhere during lowering is appropriate

Commit:
Preserve semicolon information.

This commit pivots from the previous naive solution of checking spans
towards preserving semicolon information during THIR and MIR lowering
via BlockTailInfo, BlockTailTemp, and an enum for tracking if a tail
result is ignored and whether the block has a trailing semicolon.

This commit currently isn't successful.

joculatrix commented:
I've pivoted towards preserving that info in an enum as suggested. The current commit doesn't work yet and breaks a lot of tests, so it's still a work in progress, but I think I'm close (and wanted to push my work so far just so it's clear I'm working on it).
